common/workunit/workunit.cpp

```diff
@@ -5058,6 +5058,118 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+    const char *importToIP, const char *importToPath, bool importQueryAssociatedFile)
+{
+    class CImportWorkUnitHelper : public CSimpleInterface
+    {
+        IPTree* wuTree;
+        StringAttr importToIP, importToPath;
+        StringBuffer importToPathWithIP;
+        void copyContentFromRemoteFile(const char* sourceFileName, const char* sourceIPAddress, const char *outFileName)
+        {
+            RemoteFilename rfn;
+            rfn.setRemotePath(sourceFileName);
+            SocketEndpoint ep(sourceIPAddress);
+            rfn.setIp(ep);
+
+            OwnedIFile srcFile = createIFile(rfn);
+            if (!srcFile)
+                throw MakeStringException(0, "Cannot open %s from %s.", sourceFileName, sourceIPAddress);
+
+            OwnedIFile dstFile = createIFile(outFileName);
+            if (!dstFile)
+                throw MakeStringException(0, "Cannot open %s.", outFileName);
+
+            copyFile(dstFile, srcFile);
+        }
+
+    public:
+        IMPLEMENT_IINTERFACE_USING(CSimpleInterface);
+
+        CImportWorkUnitHelper(IPTree *_wuTree, const char *_importToIP, const char *_importToPath)
+            : wuTree(_wuTree), importToIP(_importToIP), importToPath(_importToPath)
+        {
+            importToPathWithIP.set("//").append(importToIP).append(importToPath);
+        };
+
+        void resetWUProcessLogAttrs(const char* xpath)
+        { 
+            Owned<IPropertyTreeIterator> it = wuTree->getElements(xpath);
+            ForEach (*it)
+            {
+                StringBuffer name, log, newLogAttr;
+                IPropertyTree& proc = it->query();
+                proc.getName(name);
+                proc.getProp("@log", log);
+                if (name.isEmpty() || log.isEmpty())
+                    continue;
+
+                //Match the path and the name of the log files from ZAP report.
+                newLogAttr.set(importToPathWithIP).append(name).append('_').append(getFileNameOnly(log, false));
+                proc.setProp("@log", newLogAttr);
+            }
+        }
+
+        void importQueryAssociatedFiles()
+        {   //Copy Query Associated Files from original locations to the importToIP.
+            //Also update the WU xml.
+            Owned<IPropertyTreeIterator> it = wuTree->getElements("Query/Associated/File");
+            ForEach (*it)
+            {
+                IPropertyTree& file = it->query();
+                const char* sourceIP = file.queryProp("@ip");
+                const char* sourceName = file.queryProp("@filename");
+                const char pathSep = getPathSepChar(sourceName);
+                const char* namePtr = strrchr(sourceName, pathSep);
+                if (!namePtr || isEmptyString(namePtr + 1) || isEmptyString(sourceIP))
+                    continue;
+
+                StringBuffer fileNew;
+                fileNew.set(importToPath).append(namePtr + 1);
+                copyContentFromRemoteFile(sourceName, sourceIP, fileNew);
```

Comment:
Where is it copying the files from?  From the zap file?  This looks like it is trying to copy them from the location mentioned in the zap file - which won't exist.

Reply:
How about a subdir of the "temp" directory fetched from getConfigurationDirectory? Subdir name: new WUID?

The existing ZAP report does not contain the associate query files. Should I add them to ZAP report?

Reply:
You have to assume the environment has nothing.
The import of a ZAP file needs to function based on whatever is in the ZAP file only.

It might be good to make adding the associated files to the zap file an option, but a separate issue.
If they are included and restored, then the original locations are quite likely not to be relevant anymore.
(consider a zap file from one environment being imported into another).
So _if_ you were restoring associated file that were from a ZAP, it should put them somewhere (could be anywhere), and then update the locations in the workunit xml to point to them.

Reply:
>How about a subdir of the "temp" directory fetched from getConfigurationDirectory? Subdir name: new WUID?

You could, but as I mention in another comment, it might be worth looking at reading the zip file content directly, you may not need to extract to a directory at all - which would be more efficient.

Reply:
Are you saying that the WUID in the ZAP file content should be the subdir name? 
A ZAP file is a zip file. To read the content, I need to extract it to a directory.
If ZAP does not contain the associated files, but, the importQueryAssociatedFile flag is set, should I try to copy them from original locations?

Reply:
>If ZAP does not contain the associated files, but, the importQueryAssociatedFile flag is set, should I try to copy them from original locations?

No. If you focus on the scenario that you receive a ZAP file from someone and you import it to your environment.
Any existing locations are meaningless.

Reply:
>A ZAP file is a zip file. To read the content, I need to extract it to a directory.

As I say, see previous comments re. reading zip files directly - using IFile's and the libarchive installed hooks, I think you can get a director iterator and IFile's - from the zip - without extracting the files first.

Reply:
>Are you saying that the WUID in the ZAP file content should be the subdir name?

if you end up not using libarchive for some reason, and therefore have to exact to a directory, and if it's temporary, then any unique temporary directory name is good. There's probably a jfile function to get a temp directory? (if not there prob. should be).